### PR TITLE
[1.23] Improve eth flow 2

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary
Refactoring of: https://github.com/cowprotocol/cowswap/pull/888

**Cases:**
1. Native wrapped token **is approved** AND **enough native** wrapped token balance for swap
  - Expert mode ([demo](https://user-images.githubusercontent.com/7122625/194071916-017bb22a-5c95-4d6d-9ddc-de0d7515c2b3.mov))
    - Immediately send swap transaction (without confirmation)
  - Regular mode ([demo](https://user-images.githubusercontent.com/7122625/194071904-aff8c724-f2c5-4bf1-9f0d-409f828ce4be.mov))
    - show "No need to wrap ETH" modal
    - click "Swap"
    - open swap confirmation modal

2. Native wrapped token is **NOT approved** AND **enough native** wrapped token balance for swap
  - Expert mode ([demo](https://user-images.githubusercontent.com/7122625/194072128-84e37b72-7c72-4e29-816c-f3b490bfe260.mov))
    - show "Approve ETH" modal and immediately send the transaction
    - wait until tx mined
    - open swap confirmation modal
  - Regular mode ([demo](https://user-images.githubusercontent.com/7122625/194072167-ae81c9f3-a22c-4762-a4fd-9ef7ebb9318d.mov))
    - show "Approve ETH" modal
    - click "Approve ETH" and send the transaction
    - wait until tx mined
    - click "Swap"
    - open swap confirmation modal

3. Native wrapped token **is approved** AND **NOT enough native** wrapped token balance for swap
  - Expert mode ([demo](https://user-images.githubusercontent.com/7122625/194072244-1034a64a-2e03-436b-a59f-bb341388d785.mov))
    - show "Swap with Wrapped ETH" modal and immediately send the transaction
    - wait until tx mined
    - open swap confirmation modal
  - Regular mode ([demo](https://user-images.githubusercontent.com/7122625/194072317-8937fa55-b4c0-4d59-926d-2bcf59b68127.mov))
    - show "Swap with Wrapped ETH" modal
    - click "Wrap ETH" and send the transaction
    - wait until tx mined
    - show "No need to wrap ETH" modal
    - click "Swap"
    - open swap confirmation modal

4. Native wrapped token is **NOT approved** AND **NOT enough native** wrapped token balance for swap
  - Expert mode ([demo](https://user-images.githubusercontent.com/7122625/194072460-8a622d19-3e2a-4d4f-a6a0-a6d0454ca352.mov))
    - show "Wrap and approve" modal and immediately send 2 transactions
    - wait until txs mined
    - open swap confirmation modal
  - Regular mode ([demo](https://user-images.githubusercontent.com/7122625/194072483-c29a3891-1f60-4d70-bb5e-ace93ee3674d.mov))
    - show "Swap with Wrapped ETH" modal
    - click "Wrap ETH" and send the transaction
    - wait until tx mined
    - show "Approve ETH" modal
    - click "Approve ETH" and send the transaction
    - wait until tx mined
    - click "Swap"
    - open swap confirmation modal

### CODE REVIEW

File: src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts
Comment (shoom3301) [inline src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts:13]: @anxolin Please, review the config

**Comment by elena-zh** [comment]
Hey @shoom3301 , great changes!

There are some other cases we need to take into account: 
1. Wrap token:  
 - regular mode: 
   - waiting modal should appear
   - once transaction is signed, pending 'Wrap you ETH' modal should be displayed
image
   - once TX fails, it should show:
image
     - once executed, it should be closed
  - Expert mode: 
     - TX is immediately sent to the connected wallet w/o Waiting modal
     - Wrap your ETH modal is displayed at this time
     - image

   - once transaction is signed, TX log is added to the 'Wrap you ETH' modal 
image

   - once TX fails, it should show:
image
     - once executed, it should be closed

2. A similar flow should be for Unwrap transaction
unwrap


In the current PR all modals are missing when wrap/unwrap tokens.

**Comment by elena-zh** [comment]
Notes, related to the Scenario 1: 
1. Expert mode: I see ETH in the order pop-up, however, it is reported in #1125
2. Regular mode: I think, it would be great to rewrite the text, as there are 2 sentences that look similar to me
image

**Comment by elena-zh** [comment]
Notes, related to scenario 2: 
1. Expert mode: it would be great to add a period in the end of the sentence and add an explanation why I approve WETH, however, I have specified ETH in the Sell field
image
image
image

2. Expert mode: WETH icon is not displayed in the Approve ETH modal in Gorli
image
3. Expert mode: once Approve TX fails, there is no 'Approve' button on the modal to run the TX again
image
4. Expert mode: once I have approved WETH, but not enought amount for trading, I see Confirm swap modal, but, actually can't place an order as:
image
This message should be displayed:
image

5. Order Submitted modal appears in the Expert mode, but it shouldn't
image

6. Regular mode: no buttons to approve a token (and then I'm blocked in testing this scenario)
image

**Comment by elena-zh** [comment]
Notes, related to scenario 3:
1. Regular mode: there is no button on the modal to proceed:
image
2. Again, no icon for the WETH token in the modal (for Gorli network only)
image
3. There is a boarder issue I have reported recently #1151 
4. Expert mode: it would be great to add a period to the end of the sentence. Plus add some kind of a description why I need to wrap ETH
image
5. Expert mode: there are no amount that I wrap whilt the TX is pending
image
6. Expert mode: add a period to the end of the sentence
image
7. Expert mode: order submitted modal appears in the Expert mode (I reported it in the case 2)
image

**Comment by elena-zh** [comment]
Notes, related to scenario 4:
1. Regular mode: there is no button on the modal to proceed:
image
2. Expert mode: it would be great to add an explanation to the modal, plus add a period to the end of the sentence
image
3. Order of transactions is incorrect: pending one should be at the bottom, as it is the 2nd transaction
image
4. Expert mode: once one or both transactions fail, there is no way to proceed:
image
image
image

5. Expert mode: when token approval amount is not enough, the app shows Confirm order modal, but it can't be executed: 
image
**Expected**: 
image

**Review by anxolin — APPROVED** [review]
Amazing refactor, and thanks for making the effort of the videos. They help!

I will keep reading the PRs, and test it functionally mostly in the last one

**Comment by anxolin** [comment]
Issues I found is 

The button didn't show up when I tried to swap xDAI for USDC (and i didn't have enough wxDAI for the trade)

image

The expert/non-expert in wrap/unwrap (when u select xdai and wxdai) works as I expect now. They work in the same way :) 

Expert for the ETH flow, im kind of find with the current approach, although i thought it was a bit weird to change the currency. But I think we can leave it, after all, this is only for experts :) 


This has been. a great job! thanks 🙏

**Comment by shoom3301** [comment]
@anxolin fixed

File: src/cosmos.decorator.tsx
Comment (shoom3301) [reply src/cosmos.decorator.tsx:27]: As I see, the file is fully under our maintenance

File: src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts
Comment (josojo) [inline src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts:91]: Maybe you wanna update it already here?

**Review by josojo — APPROVED** [review]
Wow, really nice.

Thanks! I think the `EthFlowContext` state-object is very nicely chosen. With this state, all the state updates are much easier maintainable. I like it.

**Comment by elena-zh** [comment]
Hey @shoom3301 , I have done another round of testing, and here are issues that are not fixed: 
1. It would be great to update tests and add periods to the end of sentences to modals from my previous comments. @anxolin , could you please help with this?
2. No icon for WETH token in Gorli
image
3. For the cases 2, 4 in the Expert mode this issue is not fixed: 
image
4. 3 case, Expert mode: 'No need to wrap' modal blinks after the successful wrap before the Confirm Swap modal
5. No buttons in the modals for 2, 3, 4 cases in Expert mode when Approve/Wrap transaction fails
2 case expert mode no buttons
3 case epert fails
4 case approve fails
6. Progress bar modal appears in the Expert mode after signing transaction (however, should be fixed in #1167)
7. Order of transactions for the 4 case in the exprt mode is incorrect
image
8. 4 case, regular mode: 'Insufficient approved' modal blinks after successfully approved transaction before the 'No need to wrap' modal. 

Thanks

File: src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts
Comment (shoom3301) [reply src/custom/components/swap/EthFlow/pure/EthFlowModalContent/configs.ts:91]: Thank you!
I'm not sure about the text, I hope @anxolin will fix it

**Comment by elena-zh** [comment]
However, WXDAI icon is incorrect in this PR:
Screenshot_1
It should be like in https://github.com/cowprotocol/cowswap/pull/1152#issue-1395031158

**Comment by alfetopito** [comment]
Testing the original cases described in the PR description.

Everything working great on cases `1` and `2`.

On case `3`, I'm not sure this is just not clearly stated on the description or the behaviour is wrong.

Testing with both `expert mode` and `regular mode`.
WETH is approved.
I have 0.02+ ETH, 0.2+ WETH. 
I chose to sell 0.2 ETH - which according to my understanding should be possible, given that it'll use WETH - but it shows `insufficient ETH balance`.

Should it require ETH balance to use WETH balance?

Screen Shot 2022-10-18 at 11 31 55
Screen Shot 2022-10-18 at 11 31 35

Thinking back, it might be the same behaviour for `1` and `2`, I just didn't think about this scenario when testing them

**Comment by elena-zh** [comment]
Hey @alfetopito , great catch as it is a gap of the functionality in the current flow. 

However, I have to notice, that it works this way on Prod now
image

So I have created #1202 issue to address this enhancement in the next iteration
cc @W3stside , @josojo

**Pull request merged**
